File: backend/app/schemas/analytics.py
from typing import Dict, List, Optional, Tuple
from pydantic import BaseModel, Field


class QuickStats(BaseModel):
    total_scheduled_minutes: int
    completed_tasks_count: int
    focus_time_utilization: float = Field(ge=0.0, le=1.0)
    top_category: Optional[Dict[str, object]] = None
    streak_days_completed: int = 0


class TrendPoint(BaseModel):
    date: str  # YYYY-MM-DD
    total_scheduled_minutes: int
    focus_minutes: int
    completed_tasks_count: int


class DashboardData(BaseModel):
    category_minutes: Dict[str, int]
    trend: List[TrendPoint]
    completion_rate: float = Field(ge=0.0, le=1.0)
    focus_utilization: float = Field(ge=0.0, le=1.0)


class GoalBase(BaseModel):
    name: str
    category_id: Optional[str] = None
    target_type: str  # minutes | percentage_of_scheduled_time | task_completion_count
    target_value: int
    time_period: str  # daily | weekly | monthly
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_active: Optional[bool] = True


class GoalCreate(GoalBase):
    pass


class GoalUpdate(BaseModel):
    name: Optional[str] = None
    category_id: Optional[str] = None
    target_type: Optional[str] = None
    target_value: Optional[int] = None
    time_period: Optional[str] = None
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_active: Optional[bool] = None


class Goal(BaseModel):
    goal_id: str
    user_id: str
    name: str
    category_id: Optional[str] = None
    target_type: str
    target_value: int
    time_period: str
    start_date: Optional[str] = None
    end_date: Optional[str] = None
    is_active: bool
    created_at: str


class Period(BaseModel):
    start: str
    end: str


class GoalProgress(BaseModel):
    goal_id: str
    name: str
    target_type: str
    target_value: int
    period: Period
    achieved_value: int
    percent_complete: float
    status: str

File: frontend/src/api/analytics.ts
import axios from 'axios'

const API_BASE_URL = import.meta.env.VITE_API_BASE_URL || 'http://localhost:8000/api/v1'

const api = axios.create({
  baseURL: API_BASE_URL,
  headers: { 'Content-Type': 'application/json' },
})

api.interceptors.request.use((config) => {
  const token = localStorage.getItem('access_token')
  if (token) {
    config.headers = config.headers || {}
    config.headers['Authorization'] = `Bearer ${token}`
  }
  return config
})

export interface QuickStats {
  total_scheduled_minutes: number
  completed_tasks_count: number
  focus_time_utilization: number
  top_category?: { category_id: string; name: string; minutes: number }
  streak_days_completed: number
}

export interface TrendPoint {
  date: string
  total_scheduled_minutes: number
  focus_minutes: number
  completed_tasks_count: number
}

export interface DashboardData {
  category_minutes: Record<string, number>
  trend: TrendPoint[]
  completion_rate: number
  focus_utilization: number
}

export const analyticsApi = {
  getQuickStats: async (dateRange = '7d'): Promise<QuickStats> => {
    const { data } = await api.get('/analytics/quick-stats', { params: { date_range: dateRange } })
    return data
  },
  getDashboard: async (period = 'monthly'): Promise<DashboardData> => {
    const { data } = await api.get('/analytics/dashboard', { params: { period } })
    return data
  },
  exportCsv: async (startDate: string, endDate: string): Promise<Blob> => {
    const { data } = await api.get('/analytics/export', {
      params: { start_date: startDate, end_date: endDate },
      responseType: 'blob',
    })
    return data
  },
}

export default api
